manifests/resource/upstream.pp

```diff
@@ -18,82 +31,116 @@
 # Sample Usage:
 #  nginx::resource::upstream { 'proxypass':
 #    ensure  => present,
-#    members => [
-#      'localhost:3000',
-#      'localhost:3001',
-#      'localhost:3002',
-#    ],
+#    members => {
+#      'localhost:3001' => {
+#        server => 'localhost',
+#        port   => 3000,
+#      },
+#      'localhost:3002' => {
+#        server => 'localhost',
+#        port   => 3002,
+#      },
+#      'localhost:3003' => {
+#        server => 'localhost',
+#        port   => 3003,
+#      },
+#    },
 #  }
 #
 #  Custom config example to use ip_hash, and 20 keepalive connections
 #  create a hash with any extra custom config you want.
-#  $my_config = {
-#    'ip_hash'   => '',
-#    'keepalive' => '20',
-#  }
 #  nginx::resource::upstream { 'proxypass':
-#    ensure              => present,
-#    members => [
-#      'localhost:3000',
-#      'localhost:3001',
-#      'localhost:3002',
-#    ],
-#    upstream_cfg_prepend => $my_config,
+#    ensure    => present,
+#    members   => {
+#      'localhost:3001' => {
+#        server => 'localhost',
+#        port   => 3000,
+#      },
+#      'localhost:3002' => {
+#        server => 'localhost',
+#        port   => 3002,
+#      },
+#      'localhost:3003' => {
+#        server => 'localhost',
+#        port   => 3003,
+#      },
+#    },
+#    ip_hash   => true,
+#    keepalive => 20,
 #  }
+#
 define nginx::resource::upstream (
-  Optional[Array] $members                  = undef,
-  $members_tag                              = undef,
-  Enum['present', 'absent'] $ensure         = 'present',
-  Optional[Hash] $upstream_cfg_append       = undef,
-  Optional[Hash] $upstream_cfg_prepend      = undef,
-  $upstream_fail_timeout                    = '10s',
-  $upstream_max_fails                       = undef,
-  Enum['http', 'stream'] $upstream_context  = 'http',
+  Enum['present', 'absent']                 $ensure                 = 'present',
+  Enum['http', 'stream']                    $context                = 'http',
+  Optional[Hash]                            $members                = undef,
```

Comment:
can this default to an empty hash instead of optional? We have some guidelines for datatypes at https://voxpupuli.org/docs/#reviewing-a-module-pr

> * Are there new params with datatype Hash or Array? If possible, they should default to empty Hash/Array instead of undef. You can also enforce the datastructure like Array[String]

and:

> * If you can supply one or multiple values for an attribute it’s common practice to enforce the datatype for one value and an array of that datatype. An example for string is Variant[String[1],Array[String[1]]]. This can be used in the Puppet code as [$var].flatten()

can you also define the datatypes within the hash?

Reply:
You're right, I'll fix that.